### PR TITLE
 RTPS payload pool [9544]

### PR DESCRIPTION
**Pull request by IkerLuengo**
Documenting the new PayloadPool on the RTPS layer:

- Add the IPayloadPool interface to the API-ref
- Add the IChangePool interface to the API-ref
- Describe the functionality of the pool
- Describe how the default implementation works
- Explain how to create endpoints with a custom pool
- Provide an example using a custom pool

**note**: Keeping it as draft as tests will fail until all implementations are in place in Fast DDS.

### CODE REVIEW

**Comment by richiware** [comment]
Build status: [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Docs%20Manual/627/)

File: code/CodeTester.cpp [hunk 1]
```diff
@@ -933,16 +933,112 @@ void rtps_api_example_create_entities()
...
+    participant_attr.setName("participant");
+    RTPSParticipant* participant = RTPSDomain::createParticipant(0, participant_attr);
+
+    //RTPS_API_ENTITY_CREATE_WITH_PAYLOAD_POOL
+    // A simple payload pool that reserves and frees memory each time
+    class CustomPayloadPool : public IPayloadPool
+    {
+        bool get_payload(
+            uint32_t size,
+            CacheChange_t& cache_change) override
```
Comment (MiguelCompany) [inline code/CodeTester.cpp:960]:
```suggestion
                uint32_t size,
                CacheChange_t& cache_change) override
```

File: code/CodeTester.cpp [hunk 2]
```diff
@@ -933,16 +933,112 @@ void rtps_api_example_create_entities()
...
+        {
+            // Reserve new memory for the payload buffer
+            octet* payload = new octet[size];
+
+            // Assign the payload buffer to the CacheChange and update sizes
+            cache_change.serializedPayload.data = payload;
+            cache_change.serializedPayload.length = size;
+            cache_change.serializedPayload.max_size = size;
+
+            // Tell the CacheChange who needs to release ist payload
```
Comment (MiguelCompany) [inline code/CodeTester.cpp:970]:
```suggestion
            // Tell the CacheChange who needs to release its payload
```

File: code/CodeTester.cpp [hunk 3]
```diff
@@ -933,16 +933,112 @@ void rtps_api_example_create_entities()
...
+            cache_change.payload_owner(this);
+            
+            return true;
+        }
+
+        bool release_payload(
+                CacheChange_t& cache_change) override
+        {
+            // Dealloc the buffer of the payload
```
Comment (MiguelCompany) [inline code/CodeTester.cpp:1001]:
```suggestion
        {
            // Ensure precondition
            assert (this == cache_change.payload_owner());

            // Dealloc the buffer of the payload
```

File: code/CodeTester.cpp [hunk 4]
```diff
@@ -933,16 +933,112 @@ void rtps_api_example_create_entities()
...
+
+            // Copy the data
+            memcpy(payload, data.data, data.length);
+
+            // Assign the payload buffer to the CacheChange and update sizes
+            cache_change.serializedPayload.data = payload;
+            cache_change.serializedPayload.length = data.length;
+            cache_change.serializedPayload.max_size = data.length;
+
+            // Tell the CacheChange who needs to release ist payload
```
Comment (MiguelCompany) [inline code/CodeTester.cpp:992]:
```suggestion
            // Tell the CacheChange who needs to release its payload
```

File: code/CodeTester.cpp [hunk 5]
```diff
@@ -933,16 +933,112 @@ void rtps_api_example_create_entities()
...
+    WriterAttributes writer_attr;
+    RTPSWriter* writer = RTPSDomain::createRTPSWriter(participant, writer_attr, payload_pool, history);
+
+    // A reader using the same instance of the custom payload pool
+    HistoryAttributes history_attr;
+    ReaderHistory* history = new ReaderHistory(history_attr);
+    ReaderAttributes reader_attr;
+    RTPSReader* reader = RTPSDomain::createRTPSReader(participant, reader_attr, payload_pool, history);
+
+    // Write and Read operations work as usual, but take the Payloads from the poo.
```
Comment (MiguelCompany) [inline code/CodeTester.cpp:1030]:
```suggestion
    // Write and Read operations work as usual, but take the Payloads from the pool.
```

File: docs/fastdds/rtps_layer/rtps_layer.rst [hunk 6]
```diff
@@ -174,3 +189,102 @@ You can specify a maximum amount of changes for the History to hold and an initi
...
+Different pool implementations allow for different optimizations.
+For example, Payloads of different size could be retrieved from different preallocated memory chunks.
+
+Writers and Readers can automatically select a default Payload pool implementation that best suits
+the configuration given in |HistoryAttributes-api|.
+However, a custom Payload pool can be given to |RTPSDomain::createRTPSWriter-api| and
+|RTPSDomain::createRTPSReader-api| functions.
+Writers and Readers will use the provided pool when a new :class:`CacheChange_t` is requested
+or released.
+
```
Comment (MiguelCompany) [inline docs/fastdds/rtps_layer/rtps_layer.rst:218]: May be interesting to add a note regarding the requirements for using a custom pool in the context of standard RTPS serialization. Mentioning section 10.2 of the RTPS standard and the 4 bytes required for `representation_identifier` and `representation_options`.

File: docs/fastdds/rtps_layer/rtps_layer.rst [hunk 7]
```diff
@@ -174,3 +189,102 @@ You can specify a maximum amount of changes for the History to hold and an initi
...
+**PREALLOCATED_WITH_REALLOC_MEMORY_MODE**
+
+Payloads are guaranteed to have a data buffer at least as large as the 
+maximum between the requested size and |HistoryAttributes::payloadMaxSize-api|.
+Released Payloads can be reused for another :class:`CacheChange_t`.
+If there is at least one free Payload with a buffer size equal or larger to the requested one,
+no memory allocation is done.
+
+During the initialization of the History, |HistoryAttributes::initialReservedCaches-api|
+Payloads are preallocated for the initially allocated :class:`CacheChange_t`.
```
Comment (MiguelCompany) [inline docs/fastdds/rtps_layer/rtps_layer.rst:263]: This also applies to PREALLOCATED_MEMORY_MODE

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

File: code/CodeTester.cpp (hunk 5 above)
Comment (IkerLuengo) [reply code/CodeTester.cpp:1030]: :+1: 
that would be a shitty Payload

**Comment by richiware** [comment]
Build status: [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Docs%20Manual/628/)

**Comment by richiware** [comment]
Build status: [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Docs%20Manual/629/)

**Comment by richiware** [comment]
Build status: [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Docs%20Manual/630/)

**Comment by richiware** [comment]
Build status: [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Docs%20Manual/632/)

**Review by MiguelCompany — APPROVED** [review]

**Pull request merged**